types/encoding.go

```diff
@@ -136,8 +136,15 @@ func (d *Decoder) Read(p []byte) (int, error) {
 		if want > len(d.buf) {
 			want = len(d.buf)
 		}
-		var read int
-		read, d.err = io.ReadFull(&d.lr, d.buf[:want])
+		read, err := io.ReadFull(&d.lr, d.buf[:want])
+		if err != nil {
+			// When the decoder encounters an error, it must clear its buffer and
+			// return empty values for the remaining decodes. Because the decoder uses
+			// sticky errors instead of immediately returning a bad decode can
+			// potentially cause a massive allocation.
+			d.SetErr(err)
+			return n, err
+		}
```

Comment:
this could also be written as:
```go
read, err := io.ReadFull(&d.lr, d.buf[:want])
n += copy(p[n:], d.buf[:read])
d.SetErr(err)
```
which feels a bit more in line with the "unconditional" style of the decoder. It also handles any trailing bytes that were read (though in practice this is unlikely to matter).

hmm, if we upgrade to go 1.21 then we can make it even more concise:
```go
func (d *Decoder) Read(p []byte) (int, error) {
	n := 0
	for len(p[n:]) > 0 && d.err == nil {
		read, err := io.ReadFull(&d.lr, d.buf[:min(len(p[n:]), len(d.buf))])
		n += copy(p[n:], d.buf[:read])
		d.SetErr(err)
	}
	return n, d.err
}
```
tempting! 😅